virtual_eeprom/virtual_eeprom.c

```diff
@@ -0,0 +1,219 @@
+    if (!np) {
+        pr_err("No device node found for virtual_eeproms\n");
+        return -ENODEV;
+    }
+
+    // Iterate through all child nodes
+    for_each_child_of_node(np, child) {
+        const char *name = child->name;
+        const char *bin_path;
+        u32 size;
+
+        if (of_property_read_u32(child, "size", &size)) {
+            pr_err("Failed to read size for %s\n", name);
+            continue;
+        }
+
+        if (of_property_read_string(child, "bin-path", &bin_path)) {
+            pr_err("Failed to read bin-path for %s\n", name);
+            continue;
+        }
+
+        struct virtual_eeprom *eeprom = kzalloc(sizeof(struct virtual_eeprom), GFP_KERNEL);
+        if (!eeprom) {
+            pr_err("Failed to allocate memory for %s\n", name);
+            continue;
+        }
+
+        eeprom->data = kzalloc(size, GFP_KERNEL);
+        if (!eeprom->data) {
+            pr_err("Failed to allocate memory for EEPROM %s\n", name);
+            kfree(eeprom);
+            continue;
+        }
+
+        mutex_init(&eeprom->mutex);
+        eeprom->size = size;
+        eeprom->name = name;
+        eeprom->bin_path = bin_path;
+
+        // Create unique platform device name by appending EEPROM name
+        eeprom->pdev = platform_device_register_simple(name, -1, NULL, 0);
+        if (IS_ERR(eeprom->pdev)) {
+            pr_err("Failed to register platform device for %s\n", name);
+            kfree(eeprom->data);
+            kfree(eeprom);
+            continue;
+        }
+
+        ret = load_eeprom_from_file(eeprom);
+        if (ret < 0) {
+            // Initialize with 0 if file load fails
+            memset(eeprom->data, 0, size);
+        }
+
+        eeprom->nvmem_cfg.name = name;
+        eeprom->nvmem_cfg.type = NVMEM_TYPE_EEPROM;
+        eeprom->nvmem_cfg.read_only = false;
+        eeprom->nvmem_cfg.root_only = false;
+        eeprom->nvmem_cfg.size = size;
+        eeprom->nvmem_cfg.reg_read = eeprom_read;
+        eeprom->nvmem_cfg.reg_write = eeprom_write;
+        eeprom->nvmem_cfg.owner = THIS_MODULE;
+        eeprom->nvmem_cfg.priv = eeprom;
+        eeprom->nvmem_cfg.dev = &eeprom->pdev->dev;
+
+        eeprom->nvmem_dev = nvmem_register(&eeprom->nvmem_cfg);
+        if (IS_ERR(eeprom->nvmem_dev)) {
+            pr_err("Failed to register nvmem device for %s: %ld\n", name, PTR_ERR(eeprom->nvmem_dev));
+            platform_device_unregister(eeprom->pdev);
+            kfree(eeprom->data);
+            kfree(eeprom);
+            continue;
+        }
+
+        list_add(&eeprom->list, &eeprom_list);
+
+        pr_info("Virtual EEPROM %s loaded with nvmem\n", name);
+    }
+
+    return 0;
+}
+
+static void __exit eeprom_exit(void)
+{
+    struct virtual_eeprom *eeprom, *tmp;
+
+    // Iterate through all created EEPROMs and clean them up
+    list_for_each_entry_safe(eeprom, tmp, &eeprom_list, list) {
+        save_eeprom_to_file(eeprom);
```

Comment:
I don't think this is needed? Ideally the exit function would do as little work as possible, writing to a file seems heavy

Reply:
I'm thinking it would be better to not cache the file at all, what do you think?

Reply:
Yup, I don't really see the need for caching, especially since the whole file is written on write anyway.

Reply:
Yeah, could do.

Reply:
Done